arbor/util/span.hpp

```diff
@@ -22,24 +22,24 @@ namespace util {
 template <typename I>
 using span = range<counter<I>>;
 
-template <typename I, typename J>
+template <typename I, typename J, typename = std::enable_if_t<std::is_integral<std::common_type_t<I, J>>::value>>
 span<std::common_type_t<I, J>> make_span(I left, J right) {
     return span<std::common_type_t<I, J>>(left, right);
 }
 
 template <typename I, typename J>
 span<std::common_type_t<I, J>> make_span(std::pair<I, J> interval) {
-    return span<std::common_type_t<I, J>>(interval.first, interval.second);
+    return make_span(interval.first, interval.second);
```

Comment:
Are we certain that the type of `make_span` will match `span<std::common_type_t<I, J>>`?
And why are we not now testing for `std::is_integral`?

Same question for the uses below.

Reply:
As `span<K> = range<counter<K>>`, `K` must be both `std::integral` and `std::common_type<I, J>`. Consequently all helpers forward to `make_span(L, R)`.

Reply:
Is it necessary to have this qualification here, when `counter` itself has an `is_integral` guard? Is that guard not working somehow?

Reply:
No, the guard is fine, but I think it's better style to fail here, since then we communicate to the user of `make_span` that `common_type<A, B>` is not `integral`, rather some `C` (which happens to be `common_type<A,B>`). Makes it a bit easier to diagnose those kinds of errors.